# Inteigencia Artificial<a class="tocSkip">
## Machine Learning <a class="tocSkip">
### Ingenieria Electrónica <a class="tocSkip">
### Universidad Popular del Cesar <a class="tocSkip">
### Prof.: Jose Ramón Iglesias Gamarra - [https://github.com/joseramoniglesias/](https://github.com/joseramoniglesias/) <a class="tocSkip">
  **joseiglesias@unicesar.edu.co**

# Regresión lineal múltiple con scikitlearn

<img src="https://raw.githubusercontent.com/fhernanb/fhernanb.github.io/master/docs/logo_unal_color.png" alt="drawing" width="200"/>

Aquí se muestra un ejemplo de como usar Python para ajustar un modelo de regresión lineal múltiple.

In [4]:
import pandas as pd
import os
import numpy as np
from sklearn import linear_model
from sklearn.model_selection import train_test_split

In [5]:
os.getcwd()

'C:\\ClasesUPC_2025_2\\EL469_InteligenciaArtificial\\Clases_2025_2\\MachineLearning\\Semana_2\\Clase_4'

## Los datos

Los datos a usar en esta aplicación provienen del libro [An Introduction to Statistical Learning](http://www-bcf.usc.edu/~gareth/ISL/), la variable dependiente es `sales` de un negocio que invirtió dinero en publicidad en tres medios `TV`, `radio` y `newspaper`.

El objetivo del ejercicio es ajustar el siguiente modelo de regresión

\begin{equation} \label{eq1}
\begin{split}
sales &\sim N(\mu_i, \sigma^2) \\
\mu_i &= \beta_0 + \beta_1 \, TV + \beta_2 \, radio + \beta_3 \, newspaper
\end{split}
\end{equation}

Para leer los datos usaremos la función `read_csv` de pandas y los datos está disponibles en un archivo de Excel en [este enlace](http://faculty.marshall.usc.edu/gareth-james/ISL/) y se llaman Advertising.csv.

In [15]:
datos = pd.read_csv('Advertising.csv', index_col=0)

Para mostrar la parte inicial de los datos usamos `head()`

In [16]:
datos.head()

,TV,Radio,Newspaper,Sales
1,230.1,37.8,69.2,22.1
2,44.5,39.3,45.1,10.4
3,17.2,45.9,69.3,9.3
4,151.5,41.3,58.5,18.5
5,180.8,10.8,58.4,12.9


Para conocer la dimensión de la base de datos usamos

In [17]:
datos.shape

(200, 4)

## Ajuste del modelo

Antes de ajustar el modelo vamos a dividir el conjunto de datos en `train` y `test` usando una combinación 80%-20%. Para hacer eso usamos el siguiente código.

In [21]:
X = datos[['TV', 'Radio', 'Newspaper']]
Y = datos.Sales
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=4)

Para ver las dimensiones de los objetos usamos lo siguiente.

In [22]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(160, 3)
(160,)
(40, 3)
(40,)


Inicializamos el modelo así

In [23]:
reg = linear_model.LinearRegression()

Para ajustar el modelo usamos el método `fit`.

In [24]:
reg.fit(x_train, y_train)

LinearRegression()

El objeto `reg` tiene los atributos `coef_` e `intercept_` y se pueden utilizar así:

In [25]:
print(reg.coef_)       # Para obtener los betas
print(reg.intercept_)  # Para obtener beta0

[0.04635126 0.17768722 0.00031168]
3.05789278877425


Usando la información anterior se puede escribir el modelo ajustado así:

\begin{equation} \label{eq1}
\begin{split}
\widehat{sales} &\sim N(\hat{\mu}_i, \hat{\sigma}^2) \\
\hat{\mu_i} &= 3.058 + 0.0464 \times TV + 0.1777 \times radio + 0.0003 \times newspaper \\
\hat{\sigma} &= falta
\end{split}
\end{equation}

## Explorando los resultados del modelo

Para poner a prueba el modelo ajustado vamos a predecir los valores de ventas usando la información `x_test` que luego vamos a comparar con `y_test`.

In [26]:
sales_hat = reg.predict(x_test)
sales_hat[0:4]  # Para ver los primeros 5 valores de sales_hat

array([17.2752476 , 16.74844593, 21.97674588, 24.59448907])

Ahora vamos a ver los valores observado de `sales` para comparar con `sales_hat`:

In [27]:
y_test[0:4]  # Para ver los primeros 5 valores de y_train

12     17.4
100    17.2
129    24.7
176    27.0
Name: Sales, dtype: float64

De las dos salidas anteriores podemos ver que:
- $y_1=17.4$ mientras que $\hat{y}_1=17.2752476$,
- $y_2=17.2$ mientras que $\hat{y}_2=16.74844593$ y así sucesivamente.

Para calcular el coeficiente de correlación lineal entre $y$ e $\hat{y}$ se hace lo siguiente.

In [28]:
np.corrcoef(sales_hat, y_test)

array([[1.        , 0.95439599],
       [0.95439599, 1.        ]])

De la anterior salida podemos concluir que $\rho(y, \hat{y})=0.95439599$, esto significa que los valores de $\hat{y}$ logran ser cercanos a $y$, un característica deseable en cualquier modelo de predicción.

Para calcular el $R^2$ se hace lo siguiente.

In [29]:
reg.score(x_train, y_train)

0.8911672911176578

Para calcular el Mean Square Error entre $y$ e $\hat{y}$ se hace lo siguiente.

In [30]:
np.mean((sales_hat - y_test)**2)

2.2171510725418964

## Otras métricas

Para otras métricas que se pueden usar para analizar un modelo ajustado, [en este enlace](https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics) están varias métricas disponibles en `sklearn.metrics`.

Vamos a calcular nuevamente el $R^2$ y el $MSE$ pero ahora usando `sklearn.metrics`.

In [31]:
from sklearn.metrics import r2_score, mean_squared_error

In [32]:
print(r2_score(sales_hat, y_test))
print(mean_squared_error(sales_hat, y_test))

0.8888330040089479
2.2171510725418964


De los resultados anteriores vemos que el $R^2$ y el $MSE$ calculados manualmente fueron correctos.